A couple of fixes to make more tests pass

BenjaminBossan opened the pull request:
- Test function now checks if we have a `Pipeline.steps`-like attribute by looking at the values, not the attribute name, because some sklearn objects have this with different names.
- Support for `scipy.sparse` matrices
- Support for `sklearn.Bunch`

Locally, this brings the errors down from 62 to 49. Ready for review @adrinjalali 

**Edit**

- Added a `get_module` utility function.

BenjaminBossan committed:
A couple of fixes to make more tests pass

- Test function now checks if we have a Pipeline.steps-like attribute by
  looking at the values, not the attribute name, because some sklearn
  objects have this with different names.
- Support for scipy.sparse matrices
- Support for sklearn.Bunch

BenjaminBossan committed:
Use utility function to get module name

The reasons are:

1. Reduce a bunch of boilerplate code
2. Increase robustness: inspect.getmodule may return None

adrinjalali commented on skops/io/_general.py:
```diff
@@ -4,19 +4,13 @@
     res = {
         "__class__": obj.__class__.__name__,
-        "__module__": inspect.getmodule(type(obj)).__name__,
+        "__module__": get_module(obj),
```
since at least in one occasion, and probably other places, we get the module of `obj` instead of `type(obj)`, what do you think of doing `get_module(type(obj))`? It also reads more explicit to me.

adrinjalali commented on skops/io/_sklearn.py:
```diff
@@ -97,15 +99,23 @@ def BaseEstimator_get_instance(state, src):
 # tuples of type and function that gets the state of that type
 GET_STATE_DISPATCH_FUNCTIONS = [
     (Tree, BaseEstimator_get_state),
+    (Bunch, dict_get_state),
```
do we need this? Doesn't the dict only handle this?

BenjaminBossan replied:
Yes, right

adrinjalali commented on skops/io/tests/test_persist.py:
```diff
@@ -143,13 +174,9 @@ def assert_params_equal(params1, params2):
-            or key.endswith("__transformer_list")
-        ):
-            # we deal with a pipeline or feature union
+        if _is_steps_like(val1):
+            # Deal with Pipeline.steps, FeatureUnion.transformer_list, etc.
+            assert _is_steps_like(val2)
```
are we also checking their contents?

BenjaminBossan replied:
Yes, we transform them to dicts and below we compare the dicts.

adrinjalali commented on skops/io/tests/test_persist.py:
```diff
@@ -107,11 +108,41 @@ def _tested_estimators(type_filter=None):
         # This includes BaseEstimator since they implement __getstate__ and
         # that returns the parameters as well.
-        _assert_vals_equal(val1.__getstate__(), val2.__getstate__())
+        assert_params_equal(val1.__getstate__(), val2.__getstate__())
```
how come removing the `_`?

BenjaminBossan replied:
It's a different function vals => params

BenjaminBossan committed:
Test function works if value is nested list/tuple

BenjaminBossan committed:
Remove unused request fixture from signature

BenjaminBossan committed:
Use __dict__, not __dir__

Typo

BenjaminBossan committed:
Reviewer comments

- get_module does not automatically call type, instead caller needs to
  make it explicit
- remove unnecessary get_state entry for Bunch

BenjaminBossan reviewed (commented):
I made the change regarding the explicit type call and removed the unnecessary Bunch entry.

BenjaminBossan committed:
Actually remove Bunch entry
